beluga-tunnel/src/tunnel.rs

```diff
@@ -0,0 +1,174 @@
+                                tx_out.send(bytes::Bytes::copy_from_slice(&msg.payload)).await?;
+                            }
+                            Type::StreamStart => {
+                                let Some((websocket_in, websocket_out)) = websocket_in.take().zip(websocket_out.take())
+                                else {
+                                    return Err(Error::Service(std::io::Error::other(
+                                        "restart of the same tunnel isn't supported",
+                                    )));
+                                };
+                                stream_id = msg.stream_id;
+                                service = SshService::new(websocket_in, websocket_out).await?.into();
+                            }
+                            Type::StreamReset => {
+                                return Ok(());
+                            }
+                            Type::SessionReset => {
+                                return Ok(());
+                            }
+                            Type::ServiceIds => {
+                                // pass
+                            }
+                            Type::ConnectionStart => {
+                                // pass
+                            }
+                            Type::ConnectionReset => {
+                                return Ok(());
+                            }
+                        }
+                    }
+                }
+                data = rx_in.recv() => {
+                    let bytes = data.ok_or(Error::ServiceClosed)?;
+                    let msg = Msg {
+                        msg_type: Type::Data.into(),
+                        ignorable: false,
+                        stream_id,
+                        payload: bytes.to_vec(),
+                        ..Default::default()
+                    };
+
+                    let mut out_payload = bytes::BytesMut::new();
+                    serialize_messages(&mut out_payload, msg)?;
+                    write.send(Message::Binary(out_payload.to_vec())).await?;
+                }
+                res = async { if let Some(service) = service.take() {
+                    service.handle().await
+                } else {
+                    tokio::task::yield_now().await;
+                    Ok(())
+                } } => {
+                    res?;
+                }
+            }
+        }
+    }
+}
+
+#[derive(Debug, Deserialize, Serialize)]
+struct Notify {
+    #[serde(rename = "clientAccessToken")]
+    client_access_token: String,
+    #[serde(rename = "clientMode")]
+    client_mode: String,
+    region: String,
+    services: Vec<String>,
+}
+
+fn process_received_data(mut data: bytes::Bytes) -> Result<Vec<Msg>> {
+    let mut messages = Vec::new();
+
+    while data.remaining() > (u16::BITS / 8) as usize {
+        let len = usize::from(data.get_u16());
+        let msg = Msg::decode(&data[..len])?;
```

Comment:
Avoid panicking when the u16 is longer than what's left in `data`.
```suggestion
        let Some(raw) = data.get(..len) {
            // remote lied about the length
            break;
        };
        let msg = Msg::decode(raw)?;
```

Reply:
Can we add a few roundtrip unit tests that just do the following:
```rust
let messages = [Msg {..}];
// serialize messages
let mut serialized = BytesMut::new();
serialize_messages(&mut serialized, messages).unwrap();
// deserialize the messages
let deserialized = process_received_data(serialized).unwrap();
// check if we got back the same
assert_eq!(messages, deserialized);
```

Reply:
> Can we add a few roundtrip unit tests that just do the following:
> 
> ```rust
> let messages = [Msg {..}];
> // serialize messages
> let mut serialized = BytesMut::new();
> serialize_messages(&mut serialized, messages).unwrap();
> // deserialize the messages
> let deserialized = process_received_data(serialized).unwrap();
> // check if we got back the same
> assert_eq!(messages, deserialized);
> ```

Just a reminder for myself that it still needs to do.